### ISSUE TITLE
run diffusion-based hierarchical modeling on embeddings

### ISSUE THREAD

**Issue by smruti241** [ramanathanlab/genslm#99]
@maxzvyagin @braceal @KPHippe @hengma1001, I have used genslm for the creation of embeddings. So, I have used the code, which is already given in the genslm readme, I generated synthetic sequences too. But there is no code I found for diffusion-based hierarchical modeling. Only link is there. Can you please tell me what is the code for the creation of embeddings from nucleotide sequences and code for diffusion-based hierarchical modeling? Which script you have used for diffusion modeling?

**Comment by acadev** [comment]
hi @smruti241 , can you let me know what you are planning to do with the diffusion based model?

**Comment by smruti241** [comment]
from diffusion based model I want to take out the probability. So want to use deepspeed. Do you have any idea about this @acadev ?

**Comment by braceal** [comment]
Hi @smruti241, we have an example for how the diffusion model can be run in the readme here: https://github.com/da03/hierarchical_diffusion_LM

We don’t have a collab notebook demonstrating it yet.

Have you tried using the standard GenSLM foundation model to compute token probabilities? Also, roughly what sequence lengths are you working with? If they are less than 2048 tokens, I recommend using the base model to compute probabilities to start.

**Comment by smruti241** [comment]
@braceal can you please tell me in detail how to use the standard GenSLM foundation model to compute token probabilities? 
my sequence lengths are very large, I didnt check it yet but I am interested to know how we can compute probabilities becuase I have used colab code for computing embeddings.

**Comment by braceal** [comment]
Since our models are based directly off the hugging face implementation, you can access the output logits as shown in this documentation: https://huggingface.co/docs/transformers/main_classes/output#transformers.modeling_outputs.CausalLMOutput

Once you have the logits, you can follow some of the logic in this post to compute the probabilities:
https://discuss.huggingface.co/t/generation-probabilities-how-to-compute-probabilities-of-output-scores-for-gpt2/3175

Note: You can safely ignore the generation portions of that example. A simple adjustment to the embedding notebook should allow you to output the logits and compute the probabilities. Hope this helps!

**Comment by smruti241** [comment]
Thanks @braceal I will try this and if any issue is there I will ping you again
But I have one more query, how did you create the datasets for diffusion based hierarchical modeling? and you are using deepspeed? how did you manage to install that?
because I have installed it and I didnt find deepspeed installed then I have to install it using pip. Please explain me about that

**Comment by braceal** [comment]
- The datasets for our diffusion model are specific to SARS-CoV-2. Unless your use case involves SARS-CoV-2 sequences, I would recommend using the standard foundation models available on the Globus endpoint (instructions in readme). To answer your question, we trained the diffusion model on  >1.5 million high-quality BV-BRC SARS-CoV-2 genomes.

- Our method of installing deepspeed (and other environment parameters) is a bit specific to the clusters we used for training. Our system administrators generally provide pre-built installations that we use which are optimized to the given computing resources. In case it’s helpful, we provide installation instructions here (but  the deepspeed package is already installed in the base environments). https://github.com/ramanathanlab/genslm/blob/main/docs/INSTALL.md

- We’ve built the GenSLM interface such that you don’t need to use deepspeed (unless you want to). Our training scripts are based on PyTorch lightning which have helpful interfaces to use deepspeed. I believe installation via pip should work for this.

**Comment by smruti241** [comment]
@braceal I have some other queries about genslm. I am mentioning here:
1. Can I use code for computing embeddings which is given on github page? or any other embedding code is given? 
2. what is the length of vectors after embedding? Is it 1*768 or 2*768?
3. How can I visualize embeddings using tsne? is it available in analysis.ipynb on github page?
4. How to get output logits ? why it is needed? do we are going to get logits from embeddings? Please explain me in detail how did you use the whole genslm for computing embeddings and probabilities?

**Comment by braceal** [comment]
Hi @smruti241 

1. Yes, you can use the example code for computing embeddings. Note that the example code applies a mean pooling over the sequence length. If you need the full embeddings of shape (N_sequences, seqlength=2048, hidden_dim) then you should remove this line `emb = np.mean(emb, axis=1)`.

2. The "genslm_25M_patric" model has a hidden dimension of 512.

3. You can use t-SNE as implemented in sklearn. Generally speaking, inputting the mean-pooled embeddings from the Readme example is sufficient for visualization. https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html

4. The logits are needed to compute the token proba ilities. They are different from the embeddings, but can be accessed similarly via the hugging face API linked above.

**Comment by smruti241** [comment]
@braceal thanks for the clarification. I need average of all embeddings, so removing this line emb = np.mean(emb, axis=1) wont impact anything, right?

**Comment by smruti241** [comment]
Thanks @braceal and @KPHippe I will try all these things

**Comment by smruti241** [comment]
> * The datasets for our diffusion model are specific to SARS-CoV-2. Unless your use case involves SARS-CoV-2 sequences, I would recommend using the standard foundation models available on the Globus endpoint (instructions in readme). To answer your question, we trained the diffusion model on  >1.5 million high-quality BV-BRC SARS-CoV-2 genomes.
> * Our method of installing deepspeed (and other environment parameters) is a bit specific to the clusters we used for training. Our system administrators generally provide pre-built installations that we use which are optimized to the given computing resources. In case it’s helpful, we provide installation instructions here (but  the deepspeed package is already installed in the base environments). https://github.com/ramanathanlab/genslm/blob/main/docs/INSTALL.md
> * We’ve built the GenSLM interface such that you don’t need to use deepspeed (unless you want to). Our training scripts are based on PyTorch lightning which have helpful interfaces to use deepspeed. I believe installation via pip should work for this.

@braceal  @KPHippe Then how did you run the diffusion based hierarchical modeling and what is the inference from this?

**Comment by smruti241** [comment]
@braceal again the process is really confusing. I have some queries again
1. Genslm: how did you compute the probabilities for each sequence token and how are you assigning amino acid for each codon?
2. reward guided beam search: why are you using it and how did you generating the sequence and compute the scores
3. diffusion based hierarchical modeling: how and why are you using this? like input, data processing, etc

Please let me know about this. Thanks!

**Comment by acadev** [comment]
@smruti241 - please follow the code and most of the questions you are asking can be answered by understanding the code. The code is quite self-explanatory in terms of how we compute the probabilities from each sequence. Perhaps, it would help if you have background in language models and how to build them using PyTorch libraries? The assignment for each amino acid is based on the standard mapping for each codon -- again from basic bioinformatics look up. 

2. You probably do not need to ever use the reward guided beam search. But the description in the paper provides details on how this was implemented. Unless you are generating new sequences this would not be needed. 
3. The hierarchical diffusion model is basically used to account for the correct placement of the genes within the genome context. If you recall, the language model really models only individual genes (the causal LLM is based on learning the embeddings from ~100 million gene sequences from prokaryotes) but for the SARS-CoV-2 genomes, we need to "stitch" the correct context of how the viral genes/open reading frames are organized. So the hierarchical diffusion model provides a "jump" in that longer context (after all the entire SARS-CoV-2 genome is about 10,000 codons/ 30,000 nucleotides). The input is a GenSLM model (checkpoint) from which the hierarchical diffusion takes the SARS-CoV-2 genes (as an additional input) and tries to figure out the right order of placing them and can output an entire (hopefully valid) SARS-CoV-2 genomes.

Again, you have not responded to our original question regarding what is it that you are specifically trying to do? Both (2) and (3) are more advanced and they are really experimental at this stage of how we have implemented / built them.

**Comment by acadev** [comment]
This is not a bug report.

**Comment by smruti241** [comment]
@acadev I am trying to use hierarchical modeling as part of the paper. I just want to know why it is used and what is the input file. But my basic questions are related to genslm only. Thank you for the information you gave but the codes in the github page are pretty confusing. I am not able to understand properly what is going on because you have given code for computing embeddings and generating synthetic sequences but what is the reason behind that. Another code you have given in COMMANDS.md regarding generating embeddings, so it is really confusing which code should a user use for generating embeddings.

**Comment by smruti241** [comment]
@braceal, @KPHippe, @acadev I have computed the embeddings but the embeddings are giving me 1 class, but it should have 2 classes. Can you please tell me where I am going wrong?
